hal/src/peripherals/adc/d11/mod.rs

```diff
@@ -0,0 +1,213 @@
+use pac::adc::ctrlb::Resselselect;
+use pac::adc::inputctrl::Gainselect;
+use pac::Peripherals;
+pub mod pin;
+
+pub struct Adc0 {
+    _adc: pac::Adc,
+}
+
+impl PrimaryAdc for Adc0 {}
+
+impl AdcInstance for Adc0 {
+    type Instance = pac::Adc;
+    type Clock = crate::clock::AdcClock;
+
+    #[cfg(feature = "async")]
+    type Interrupt = crate::async_hal::interrupts::ADC;
+
+    #[inline]
+    fn peripheral_reg_block(p: &mut Peripherals) -> &pac::adc::RegisterBlock {
+        &p.adc
+    }
+
+    #[inline]
+    fn enable_pm(pm: &mut pac::Pm) {
+        pm.apbcmask().modify(|_, w| w.adc_().set_bit());
+    }
+
+    #[inline]
+    fn calibrate(_instance: &Self::Instance) {}
+
+    #[cfg(feature = "async")]
+    #[inline]
+    fn waker() -> &'static embassy_sync::waitqueue::AtomicWaker {
+        use super::async_api;
+        &async_api::ADC_WAKERS[0]
+    }
+}
+
+impl<I: AdcInstance> Adc<I, NoneT> {
+    #[inline]
+    pub fn configure(&mut self, config: Config) -> Result<(), super::Error> {
+        // Reset ADC here as we cannot guarantee its state
+        // This also disables the ADC
+        self.software_reset();
+        I::calibrate(&self.adc);
+        self.sync();
+        self.adc
+            .ctrlb()
+            .modify(|_, w| w.prescaler().variant(config.clk_divider));
+        self.sync();
+        self.adc
+            .ctrlb()
+            .modify(|_, w| w.ressel().variant(config.bit_width));
+        self.sync();
+
+        self.adc
+            .sampctrl()
+            .modify(|_, w| unsafe { w.samplen().bits(config.sample_clock_cycles) }); // sample length
+        self.sync();
+        self.adc.inputctrl().modify(|_, w| {
+            w.muxneg().gnd();
+            w.gain().variant(Gainselect::Div2)
+        }); // No negative input (internal gnd)
+        self.sync();
+
+        // Check bit width selected
+        if config.accumulation != AdcAccumulation::Single
+            && config.bit_width != Resselselect::_16bit
+        {
+            return Err(super::Error::InvalidSampleBitWidth);
+        }
+        match config.accumulation {
+            AdcAccumulation::Single => {
+                // 1 sample to be used as is
+                self.adc.avgctrl().modify(|_, w| {
+                    w.samplenum().variant(Samplenumselect::_1);
+                    unsafe { w.adjres().bits(0) }
+                });
+            }
+            AdcAccumulation::Average(adc_sample_count) => {
+                // A total of `adc_sample_count` elements will be averaged by the ADC
+                // before it returns the result
+                self.adc.avgctrl().modify(|_, w| {
+                    w.samplenum().variant(adc_sample_count);
+                    unsafe {
+                        // Table 45-3 SAME51 datasheet
+                        w.adjres()
+                            .bits(core::cmp::min(adc_sample_count as u8, 0x04))
+                    }
+                });
+            }
```

Comment:
I think the ADJRES math needs to be reviewed here... The bits define the division coefficient in 2^n increments. For example, with a sample count of 4, the result would get divided by 16. I doubt that's actually the intention here?

Reply:
Well, it appears I was looking at the wrong table after all. After testing on hardware it indeed does appear to work as intended.